drop support for Node 12

warner opened the pull request:
This removes the CI jobs that tested everything against Node 12, leaving only Node 14. It updates the top-level README to indicate the new requirement, and sets the top-level `package.json` to require `node: ">=14.15.0"`, which covers the 14 LTS releases.

It also changes all the individual package `package.json` to require `>= 14.15.0`. I'm less confident about this change. It's in a separate commit to make it easy to remove if reviewers say so.

mhofman commented on README.md:
```diff
@@ -17,7 +17,8 @@ to use.
 * Git
-* Node.js (version 12.14.1 or higher)
+* Node.js LTS (version 14.15.0 or higher)
+  * we generally support the latest LTS release: use [nvm](https://github.com/creationix/nvm.git) to keep your local system up-to-date
```
```suggestion
  * we generally support the latest LTS release: use [nvm](https://github.com/nvm-sh/nvm) to keep your local system up-to-date
```

warner commented:
cc @tyg , I don't know where in our docs tree we talk about minimum Node.js versions (maybe the Getting Started page or something just before that?), but it will need to be updated too.

mhofman commented on package.json:
```diff
@@ -63,7 +63,7 @@
   },
   "engines": {
-    "node": ">=12.14.1"
+    "node": ">=14.15.0"
```
If we don't enforce individual packages, should we enforce the top level? I suppose it's never `npm install`ed anywhere directly, but wondering what the behavior of `yarn install` is.

mhofman replied:
Doubled checked in a container running node 12 and `yarn install` fails, so this change should move to the second commit (or be reverted if we decide not to include the second commit)

warner replied:
I think we want `yarn install` to fail if the user's Node is too old (fail fast), so I'm pretty sure we want the top-level enforcement.

The purpose of the bottom-level `package.json` changes would be to make a similar enforcement on individual packages if e.g. someone installed one from NPM, rather than using our monorepo. I know we want this for at least swingset (where I'll be making a 14.x-requiring change soon). What I don't know is if it's better to have all the packages list the same requirement (yay consistency), or have each only list the requirement we know is true (yay tolerance of individual packages being used via NPM on older systems).

If we choose the former, I'll include the second commit. If we choose the latter, I'd change that commit to only update swingset (or, as @michaelfig suggested, move that `package.json` change into my upcoming PR that actually depends on WeakRef). I don't want to exhaustively test each package against different versions to see what it really requires, so I'd simply assume that the current individual requirements are suffciently correct and let them stand.

So I guess my real question is: is consistency of the requirements important/useful?

mhofman replied:
Thanks for the detailed reasoning.

Given that we currently publish individual packages to NPM, I think consistency is not as valuable. We can enforce consistency for development by keeping the root `package.json` to the latest engine version required by any of the required sub packages.

Technically that would mean dropping the 2nd commit and only requiring whatever latest is in this PR, before bumping both the SwingSet and root package.json in the next PR.

warner replied:
> Technically that would mean dropping the 2nd commit and only requiring whatever latest is in this PR, before bumping both the SwingSet and root package.json in the next PR.

Ok, are we comfortable with a commit which drops CI coverage for 12, but leaves the top-level package.json claiming `node >=12.14.1`?

mhofman replied:
Yeah I guess we should ignore the PR layering here and just require 14 for the top level.

warner commented:
I've replaced that branch with one that leaves the individual package.jsons alone, in a single commit.

dckc reviewed (changes requested):
I have a hard time seeing this as a fix; it looks more like a breaking change.

[Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) calls such things `refactor!` in their example:

```
refactor!: drop support for Node 6

BREAKING CHANGE: refactor to use JavaScript features not available in Node 6.
```

I suppose `fix` vs. `refactor` isn't critical, but the breaking change indicator (`!` and/or `BREAKING CHANGE:` footer) seems cost-effective.

warner commented:
> I have a hard time seeing this as a fix; it looks more like a breaking change.
> 
> [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) calls such things `refactor!` in their example:
> 
> ```
> refactor!: drop support for Node 6
> 
> BREAKING CHANGE: refactor to use JavaScript features not available in Node 6.
> ```
> 
> I suppose `fix` vs. `refactor` isn't critical, but the breaking change indicator (`!` and/or `BREAKING CHANGE:` footer) seems cost-effective.

Done, changed to `fix!` and added a more capital letters.

dckc reviewed (approved):
:+1:

warner commented:
Oh, huh, CI didn't finish because it's waiting for the 12.x tests that were removed. I wonder how we're supposed to change that.

warner commented:
Ah, right there are branch protection rules in the repo Settings which name the status checks that are required. I'll change those too.
